libraries/lib-audio-io/AudioIO.cpp

```diff
@@ -2015,104 +2023,208 @@ bool AudioIO::ProcessPlaybackSlices(
          // warping
          if (frames > 0) {
             size_t produced = 0;
+            // Check for asynchronous user changes in mute, solo status
+            const auto discardable = SequenceShouldBeSilent(*mPlaybackSequences[iSequence]);
             if (toProduce)
                produced = mixer->Process(toProduce);
+
             //wxASSERT(produced <= toProduce);
             // Copy (non-interleaved) mixer outputs to one or more ring buffers
-            const auto nChannels = mPlaybackSequences[iSequence++]->NChannels();
+            const auto nChannels = mPlaybackSequences[iSequence]->NChannels();
+            
             for (size_t j = 0; j < nChannels; ++j) {
-               auto warpedSamples = mixer->GetBuffer(j);
-               const auto put = mPlaybackBuffers[iBuffer++]->Put(
-                  warpedSamples, floatSample, produced, frames - produced);
-               // wxASSERT(put == frames);
-               // but we can't assert in this thread
-               wxUnusedVar(put);
+
+               auto& buffer = mProcessingBuffers[iBuffer++];
+               const auto offset = buffer.size();
+               //there could be leftovers from the previous pass, don't discard them
+               buffer.resize(buffer.size() + frames, 0);
+
+               if(!discardable)
+               {
+                  auto warpedSamples = mixer->GetBuffer(j);
+                  std::copy_n(
+                     reinterpret_cast<const float*>(warpedSamples),
+                     produced,
+                     buffer.data() + offset);
+                  std::fill_n(
+                     buffer.data() + offset + produced,
+                     frames - produced,
+                     .0f);
+               }
+               else
+                  //TODO: mixer->Process() may be wasting CPU cycles
```

Comment:
I don't think the mixer can be properly eliminated in the case that it does resampling of a track whose rate does not match the device output rate, or there is a time track.  There will be some state in the resampler dependent on previous input.

However, note that the mixer stands as a producer in relation to this function which is its consumer.  Perhaps, if it really matters, another thread could mix in advance.

But all that effort at overlapping computation may be unnecessary because this thread really spends most of its time idle anyway, going faster than the ultimate real-time consumption of input.

Or then again maybe it might in future because we want to have ever more capacity for complicated calculations like more effects, more stretching, more pitch shifting.

Or again, maybe you were thinking of power consumption, not performance.

Reply:
I though that mixer could be "advanced" without doing any actual work